rpcs3/Emu/Cell/Modules/cellSpurs.cpp

```diff
@@ -4087,10 +4092,49 @@ s32 cellSpursJobHeaderSetJobbin2Param()
 	return CELL_OK;
 }
 
-s32 cellSpursAddUrgentCommand()
+s32 cellSpursAddUrgentCommand(ppu_thread& ppu, vm::ptr<CellSpursJobChain> jobChain, u64 newCmd)
 {
-	UNIMPLEMENTED_FUNC(cellSpurs);
-	return CELL_OK;
+	cellSpurs.trace("cellSpursAddUrgentCommand(jobChain=*0x%x, newCmd=0x%llx)", jobChain, newCmd);
+
+	if (!jobChain)
+		return CELL_SPURS_JOB_ERROR_NULL_POINTER;
+
+	if (!jobChain.aligned(128))
+		return CELL_SPURS_JOB_ERROR_ALIGN;
+
+	if (jobChain->workloadId >= 32)
+		return CELL_SPURS_JOB_ERROR_INVAL;
+
+	while (true)
+	{
+		for (u32 i = 0;; i++)
+		{
+			if (i >= std::size(jobChain->urgentCmds))
+			{
+				// Exausted all slots
+				return CELL_SPURS_JOB_ERROR_BUSY;
+			}
+
+			u64 currCmd = ppu_ldarx(ppu, jobChain.ptr(&CellSpursJobChain::urgentCmds, i).addr());
+			std::atomic_thread_fence(std::memory_order_acq_rel);
+			bool found = false;
+
+			if (!currCmd)
+			{
+				if (i != 0 && !jobChain->urgentCmds[i - 1])
+					break; // Restart search, someone emptied out the previous one
+
+				found = true;
+				currCmd = newCmd;
+			}
+
+			if (!ppu_stdcx(ppu, jobChain.ptr(&CellSpursJobChain::urgentCmds, i).addr(), currCmd))
+				break; // Someone modified the job chain, restart search
+
+			if (found)
+				return CELL_OK;
```

Comment:
Maybe just use found as the condition for the while loop and simply return cell ok at the end while breaking here

Reply:
It doesnt handle ppu_stdcx failures well though.

Reply:
Can Just set found to false on failure.
Anyway. It's just for readability.
Currently it looks like the function may be void

Reply:
Maybe it's even a warning. Haven't looked at the build log yet

Reply:
I changed it to return a value at the end of function but I believe all compilers are smart enough to realise all paths of a loop either return a value or continue being in the loop. If they won't realise it they would have emitted invalid or dead code, it also applies for warnings.